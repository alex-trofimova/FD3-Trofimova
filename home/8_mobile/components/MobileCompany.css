.MobileCompany
{
  width: 800px;
  padding: 3px 7px;
}

.MobileCompanyName
{
  margin: 5px 0 10px 0;
}

.MobileCompanyClients
{
  padding-top: 5px;
}

.MobileCompanyClientsList {
  width: 100%;
  margin: 10px 0;
}

.TableHeader {
  border-bottom: 2px solid grey;
  text-align: left;
}


