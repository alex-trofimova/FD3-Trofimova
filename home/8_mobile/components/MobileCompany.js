import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';
import AddOrEditClient from './AddOrEditClient';

import {mobileCompanyEvents} from './events';

import './MobileCompany.css';

class MobileCompany extends React.PureComponent {

  static propTypes = {
    name: PropTypes.string.isRequired,//имя мобильной компании
    clients:PropTypes.arrayOf( //массив данных о ее клиентах
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
        status: PropTypes.string.isRequired,
      })
    ),
  };
  
  

  state = {
    name: this.props.name, 
    clients: this.props.clients,

    //условие в каком режиме: просмотра (=0) редактирования (=1) или добавления (=2)  
    //будет отображаться страница
    usedRegime: 0, //первоначально режим просмотра

    //хэш для добавления нового клиента, изначально пустой
    addingClient: {
      id: 0, 
      fam: '', im: '', otch:'',
      balance: 0, status: '',
    },

    //хэш для редактирования выбранного клиента, изначально пустой
    editingClient:{},

  };

  //прием событий из потока событий mobileCompanyEvents
  componentDidMount = () => {
    mobileCompanyEvents.addListener('EClientToDelete',this.clientToRemove);
    mobileCompanyEvents.addListener('EClientToEdit',this.clientToEdit);
    mobileCompanyEvents.addListener('EClientToAdd',this.clientToAdd);
    mobileCompanyEvents.addListener('EClientToCancel',this.cancel);
    mobileCompanyEvents.addListener('EClientToSaveChanges',this.clientToSaveChanges);
    
  };

  //"отписка" от них
  componentWillUnmount = () => {
    mobileCompanyEvents.removeListener('EClientToDelete',this.clientToRemove);
    mobileCompanyEvents.removeListener('EClientToEdit',this.clientToEdit);
    mobileCompanyEvents.removeListener('EClientToAdd',this.clientToAdd);
    mobileCompanyEvents.removeListener('EClientToCancel',this.cancel);
    mobileCompanyEvents.removeListener('EClientToSaveChanges',this.clientToSaveChanges);
  };

  //изменение имени компании по кнопке
  setName1 = () => {
    this.setState({name:'MTS'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };

  //описание функции для отображения всех клиентов
  showAllClients = () => {
    let newClients=this.props.clients; //создаю новый массив и кладу в него пропс - первоначальный массив клиентов
    this.setState( {clients:newClients} );//для перерендера MobileCompany в стейт ссылку на новый массив
  }

  //описание функции для отображения только активных клиентов
  //использую метод filter - он сам создает новый массив, поэтому MobileCompany перерендерится
  showActiveClients = () => {
    this.setState( {clients:this.state.clients.filter(elem => elem.status==="active")} );
  }

  //аналогично
  showBlockedClients = () => {
    this.setState( {clients:this.state.clients.filter(elem => elem.status==="blocked")} );
  }

  //описание функции отображения карточки добавления клиента при нажатии кнопки "Добавить нового клиента"
  showAddClientCard = () => {
    //предварительно составляю массив из "использованных" id
    let usedId=[];
    let newClients=[...this.state.clients];
    newClients.forEach( (elem, index) => 
      usedId[index]=elem.id
    );
    //переменная с id добавляемого клиента
    let newClientId;
    do {
      //функция выбора случайного числа из некоторого диапазон [n,m]
      let getId=(m,n)=>Math.floor(Math.random()*(m-n+1))+n;

      newClientId=getId(100,130);//здесь указываю в каком диапазоне (для проверки делала узким)
    }
    while ( usedId.indexOf(newClientId)!=-1 );

    let newAddingClient={...this.state.addingClient};
    newAddingClient.id=newClientId;
    this.setState( {usedRegime:2, addingClient:newAddingClient} );
  }

  //описание функции для удаления клиента по событию EClientToDelete
  clientToRemove = (id) => {
    this.setState( {clients:this.state.clients.filter(elem => elem.id!=id)} );       
  }

   //описание функции для редактирования клиента по событию EClientToEdit
  clientToEdit = (info) => {
    this.setState( {editingClient:info, usedRegime:1} );      
  };

  //описание функции для добавления в список нового клиента по событию EClientToAdd
  clientToAdd = (hash) => {
    let newClients=[...this.state.clients];
    newClients.push(hash);
    this.setState({clients:newClients, usedRegime:0});
        
  };
  
  //описание функции для отмены действий по событию EClientToCancel
  cancel = () => {
    this.setState( {usedRegime:0} );
  };

  //описание функции для сохранения изменений редактируемого клиента по событию EClientToSaveChanges
  clientToSaveChanges = (hash) => {
    let newClients=[...this.state.clients];
    let edIndex=newClients.findIndex(elem => elem.id==hash.id);
    newClients.splice(edIndex,1,hash);
    this.setState({clients:newClients, usedRegime:0});
  };
  
  
  render() {

    console.log("MobileCompany render");

    var clientsCode=this.state.clients.map( client => {
        return <MobileClient key={client.id} info={client}/>;
      }
    );

    return (
      <div className='MobileCompany'>
        <input type="button" value="Velcom" onClick={this.setName2} />
        <input type="button" value="MTS" onClick={this.setName1} />
        <div className='MobileCompanyName'>Компания: {this.state.name}</div>
        
        <div>
          <button className='MobileCompanyFilter_btn' value='all' 
          onClick={this.showAllClients}
          >Все</button>
          <button className='MobileCompanyFilter_btn' value='active' 
          onClick={this.showActiveClients}
          >Активные</button>
          <button className='MobileCompanyFilter_btn' value='blocked' 
          onClick={this.showBlockedClients}
          >Заблокированные</button>
        </div>

        <table className='MobileCompanyClientsList'>
          <thead className='HeaderContent'>
            <tr className='TableHeader'>
              <th className='HeaderCell'>Фамилия</th>
              <th className='HeaderCell'>Имя</th>
              <th className='HeaderCell'>Отчество</th>
              <th className='HeaderCell'>Баланс</th>
              <th className='HeaderCell'>Статус</th>
              <th className='HeaderCell'>Редактировать</th>
              <th className='HeaderCell'>Удалить</th>
            </tr>        
          </thead> 
            <tbody className='ClientsListContent'>{clientsCode}</tbody>
        </table>
        {
          (this.state.usedRegime==0) 
          ?
          <button className='MobileCompanyNewClient_btn' value='add' onClick={this.showAddClientCard}>
            Добавить клиента
          </button>
          :
          <AddOrEditClient 
            editingOrNewClient={(this.state.usedRegime==1)?this.state.editingClient:this.state.addingClient}
            usedRegime={this.state.usedRegime}
          />
        }
      </div>

    )
    ;

  }

}

export default MobileCompany;
