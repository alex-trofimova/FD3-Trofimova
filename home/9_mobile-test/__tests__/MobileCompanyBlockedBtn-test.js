import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';
let clientsArr=require('./json/clientsList.json');
let clientsArrBlocked=require('./json/clientsListBlocked.json');

let companyName='Velcom';

test('работа MobileCompany BlockedBtn', () => {

// рендер компонента
const component = renderer.create(
  <MobileCompany 
  name={companyName}
  clients={clientsArr}
  />
);

// проверка утверждений
let componentTree=component.toJSON();
expect(componentTree).toMatchSnapshot();

// найдём в вёрстке компонента саму кнопку
const buttonElem = component.root.find(  el => el.props.className == 'MobileCompanyFilter_btn BlockedBtn' );
// и "нажмём" на неё
buttonElem.props.onClick();

// обновление с некоторыми отличающимися пропсами
component.update(<MobileCompany name={companyName}
clients={clientsArrBlocked}/>);

// проверка утверждений
componentTree=component.toJSON();
expect(componentTree).toMatchSnapshot();

    
});
