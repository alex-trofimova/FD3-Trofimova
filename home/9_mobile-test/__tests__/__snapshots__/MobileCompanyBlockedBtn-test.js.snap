// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа MobileCompany BlockedBtn 1`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Velcom"
  />
  <input
    onClick={[Function]}
    type="button"
    value="MTS"
  />
  <div
    className="MobileCompanyName"
  >
    Компания: 
    Velcom
  </div>
  <div>
    <button
      className="MobileCompanyFilter_btn AllBtn"
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      className="MobileCompanyFilter_btn ActiveBtn"
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      className="MobileCompanyFilter_btn BlockedBtn"
      onClick={[Function]}
      value="blocked"
    >
      Заблокированные
    </button>
  </div>
  <table
    className="MobileCompanyClientsList"
  >
    <thead
      className="HeaderContent"
    >
      <tr
        className="TableHeader"
      >
        <th
          className="HeaderCell"
        >
          Фамилия
        </th>
        <th
          className="HeaderCell"
        >
          Имя
        </th>
        <th
          className="HeaderCell"
        >
          Отчество
        </th>
        <th
          className="HeaderCell"
        >
          Баланс
        </th>
        <th
          className="HeaderCell"
        >
          Статус
        </th>
        <th
          className="HeaderCell"
        >
          Редактировать
        </th>
        <th
          className="HeaderCell"
        >
          Удалить
        </th>
      </tr>
    </thead>
    <tbody
      className="ClientsListContent"
    >
      <tr
        className="MobileClient"
      >
        <td
          className="ClientCell ClientFam"
        >
          Иванов
        </td>
        <td
          className="ClientCell ClientIm"
        >
          Иван
        </td>
        <td
          className="ClientCell ClientOtch"
        >
          Иванович
        </td>
        <td
          className="ClientCell ClientBalance"
        >
          200
        </td>
        <td
          className="ClientCell ClientStatus"
        >
          active
        </td>
        <td
          className="ClientCell ClientEdit"
        >
          <button
            className="ClientBtn EditBtn"
            onClick={[Function]}
          >
            редактировать
          </button>
        </td>
        <td
          className="ClientCell ClientDelete"
        >
          <button
            className="ClientBtn DeleteBtn"
            onClick={[Function]}
          >
            удалить
          </button>
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td
          className="ClientCell ClientFam"
        >
          Сидоров
        </td>
        <td
          className="ClientCell ClientIm"
        >
          Сидор
        </td>
        <td
          className="ClientCell ClientOtch"
        >
          Сидорович
        </td>
        <td
          className="ClientCell ClientBalance"
        >
          250
        </td>
        <td
          className="ClientCell ClientStatus"
        >
          active
        </td>
        <td
          className="ClientCell ClientEdit"
        >
          <button
            className="ClientBtn EditBtn"
            onClick={[Function]}
          >
            редактировать
          </button>
        </td>
        <td
          className="ClientCell ClientDelete"
        >
          <button
            className="ClientBtn DeleteBtn"
            onClick={[Function]}
          >
            удалить
          </button>
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td
          className="ClientCell ClientFam"
        >
          Петров
        </td>
        <td
          className="ClientCell ClientIm"
        >
          Пётр
        </td>
        <td
          className="ClientCell ClientOtch"
        >
          Петрович
        </td>
        <td
          className="ClientCell ClientBalance"
        >
          180
        </td>
        <td
          className="ClientCell ClientStatus"
        >
          active
        </td>
        <td
          className="ClientCell ClientEdit"
        >
          <button
            className="ClientBtn EditBtn"
            onClick={[Function]}
          >
            редактировать
          </button>
        </td>
        <td
          className="ClientCell ClientDelete"
        >
          <button
            className="ClientBtn DeleteBtn"
            onClick={[Function]}
          >
            удалить
          </button>
        </td>
      </tr>
      <tr
        className="MobileClient"
      >
        <td
          className="ClientCell ClientFam"
        >
          Григорьев
        </td>
        <td
          className="ClientCell ClientIm"
        >
          Григорий
        </td>
        <td
          className="ClientCell ClientOtch"
        >
          Григорьевич
        </td>
        <td
          className="ClientCell ClientBalance"
        >
          -220
        </td>
        <td
          className="ClientCell ClientStatus"
        >
          blocked
        </td>
        <td
          className="ClientCell ClientEdit"
        >
          <button
            className="ClientBtn EditBtn"
            onClick={[Function]}
          >
            редактировать
          </button>
        </td>
        <td
          className="ClientCell ClientDelete"
        >
          <button
            className="ClientBtn DeleteBtn"
            onClick={[Function]}
          >
            удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="MobileCompanyNewClient_btn AddNewClientBtn"
    onClick={[Function]}
    value="add"
  >
    Добавить клиента
  </button>
</div>
`;

exports[`работа MobileCompany BlockedBtn 2`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Velcom"
  />
  <input
    onClick={[Function]}
    type="button"
    value="MTS"
  />
  <div
    className="MobileCompanyName"
  >
    Компания: 
    Velcom
  </div>
  <div>
    <button
      className="MobileCompanyFilter_btn AllBtn"
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      className="MobileCompanyFilter_btn ActiveBtn"
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      className="MobileCompanyFilter_btn BlockedBtn"
      onClick={[Function]}
      value="blocked"
    >
      Заблокированные
    </button>
  </div>
  <table
    className="MobileCompanyClientsList"
  >
    <thead
      className="HeaderContent"
    >
      <tr
        className="TableHeader"
      >
        <th
          className="HeaderCell"
        >
          Фамилия
        </th>
        <th
          className="HeaderCell"
        >
          Имя
        </th>
        <th
          className="HeaderCell"
        >
          Отчество
        </th>
        <th
          className="HeaderCell"
        >
          Баланс
        </th>
        <th
          className="HeaderCell"
        >
          Статус
        </th>
        <th
          className="HeaderCell"
        >
          Редактировать
        </th>
        <th
          className="HeaderCell"
        >
          Удалить
        </th>
      </tr>
    </thead>
    <tbody
      className="ClientsListContent"
    >
      <tr
        className="MobileClient"
      >
        <td
          className="ClientCell ClientFam"
        >
          Григорьев
        </td>
        <td
          className="ClientCell ClientIm"
        >
          Григорий
        </td>
        <td
          className="ClientCell ClientOtch"
        >
          Григорьевич
        </td>
        <td
          className="ClientCell ClientBalance"
        >
          -220
        </td>
        <td
          className="ClientCell ClientStatus"
        >
          blocked
        </td>
        <td
          className="ClientCell ClientEdit"
        >
          <button
            className="ClientBtn EditBtn"
            onClick={[Function]}
          >
            редактировать
          </button>
        </td>
        <td
          className="ClientCell ClientDelete"
        >
          <button
            className="ClientBtn DeleteBtn"
            onClick={[Function]}
          >
            удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="MobileCompanyNewClient_btn AddNewClientBtn"
    onClick={[Function]}
    value="add"
  >
    Добавить клиента
  </button>
</div>
`;
